File: frontend/src/store/modules/software.js
import * as softwareApi from '@/api/software'

const state = {
  spaces: [],
  currentSpace: null,
  versions: [],
  currentVersion: null,
  loading: false,
  error: null,
}

const getters = {
  spaces: state => state.spaces,
  currentSpace: state => state.currentSpace,
  versions: state => state.versions,
  currentVersion: state => state.currentVersion,
  loading: state => state.loading,
  error: state => state.error,

  // 根据ID获取软件空间
  spaceById: state => id => state.spaces.find(space => space.id === id),

  // 根据ID获取软件版本
  versionById: state => id => state.versions.find(version => version.id === id),

  // 获取当前空间的最新版本
  latestVersion: state => {
    if (!state.versions.length) return null
    return state.versions.reduce((latest, version) => {
      return new Date(version.created_at) > new Date(latest.created_at) ? version : latest
    })
  },

  // 获取已发布的版本
  publishedVersions: state => state.versions.filter(version => version.is_published),
}

const mutations = {
  SET_SPACES(state, spaces) {
    state.spaces = spaces
  },

  ADD_SPACE(state, space) {
    state.spaces.unshift(space)
  },

  UPDATE_SPACE(state, updatedSpace) {
    const index = state.spaces.findIndex(space => space.id === updatedSpace.id)
    if (index !== -1) {
      state.spaces.splice(index, 1, updatedSpace)
    }
  },

  DELETE_SPACE(state, spaceId) {
    state.spaces = state.spaces.filter(space => space.id !== spaceId)
  },

  SET_CURRENT_SPACE(state, space) {
    state.currentSpace = space
  },

  SET_VERSIONS(state, versions) {
    state.versions = versions
  },

  ADD_VERSION(state, version) {
    state.versions.unshift(version)
  },

  UPDATE_VERSION(state, updatedVersion) {
    const index = state.versions.findIndex(version => version.id === updatedVersion.id)
    if (index !== -1) {
      state.versions.splice(index, 1, updatedVersion)
    }
  },

  DELETE_VERSION(state, versionId) {
    state.versions = state.versions.filter(version => version.id !== versionId)
  },

  SET_CURRENT_VERSION(state, version) {
    state.currentVersion = version
  },

  SET_LOADING(state, loading) {
    state.loading = loading
  },

  SET_ERROR(state, error) {
    state.error = error
  },

  CLEAR_ERROR(state) {
    state.error = null
  },

  CLEAR_STATE(state) {
    state.spaces = []
    state.currentSpace = null
    state.versions = []
    state.currentVersion = null
    state.error = null
  },
}

const actions = {
  // 获取所有软件空间
  async getSpaces({ commit }) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.getSpaces()
      commit('SET_SPACES', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '获取软件空间列表失败')
      throw error
    }
  },

  // 创建软件空间
  async createSpace({ commit }, spaceData) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.createSpace(spaceData)
      commit('ADD_SPACE', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '创建软件空间失败')
      throw error
    }
  },

  // 获取软件空间详情
  async getSpace({ commit }, spaceId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.getSpace(spaceId)
      commit('SET_CURRENT_SPACE', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '获取软件空间详情失败')
      throw error
    }
  },

  // 更新软件空间
  async updateSpace({ commit }, { spaceId, spaceData }) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.updateSpace(spaceId, spaceData)
      commit('UPDATE_SPACE', response.data)

      // 如果当前空间是更新后的空间，也更新当前空间
      if (state.currentSpace && state.currentSpace.id === response.data.id) {
        commit('SET_CURRENT_SPACE', response.data)
      }

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '更新软件空间失败')
      throw error
    }
  },

  // 删除软件空间
  async deleteSpace({ commit }, spaceId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      await softwareApi.deleteSpace(spaceId)
      commit('DELETE_SPACE', spaceId)

      // 如果删除的是当前空间，清空当前空间
      if (state.currentSpace && state.currentSpace.id === spaceId) {
        commit('SET_CURRENT_SPACE', null)
      }

      commit('SET_LOADING', false)
      return { success: true }
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '删除软件空间失败')
      throw error
    }
  },

  // 重新生成API密钥
  async regenerateApiKey({ commit }, spaceId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.regenerateApiKey(spaceId)

      // 更新软件空间信息
      const space = state.spaces.find(s => s.id === spaceId)
      if (space) {
        const updatedSpace = { ...space, api_key: response.data.api_key }
        commit('UPDATE_SPACE', updatedSpace)

        if (state.currentSpace && state.currentSpace.id === spaceId) {
          commit('SET_CURRENT_SPACE', updatedSpace)
        }
      }

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '重新生成API密钥失败')
      throw error
    }
  },

  // 获取软件版本列表
  async getVersions({ commit }, spaceId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.getVersions(spaceId)
      commit('SET_VERSIONS', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '获取软件版本列表失败')
      throw error
    }
  },

  // 创建软件版本
  async createVersion({ commit }, { spaceId, versionData }) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      // 创建FormData用于文件上传
      const formData = new FormData()
      formData.append('file', versionData.file)
      formData.append('version', versionData.version)

      if (versionData.release_note) {
        formData.append('release_note', versionData.release_note)
      }

      if (versionData.documentation_url) {
        formData.append('documentation_url', versionData.documentation_url)
      }

      const response = await softwareApi.createVersion(spaceId, formData)
      commit('ADD_VERSION', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '创建软件版本失败')
      throw error
    }
  },

  // 获取软件版本详情
  async getVersion({ commit }, versionId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.getVersion(versionId)
      commit('SET_CURRENT_VERSION', response.data)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '获取软件版本详情失败')
      throw error
    }
  },

  // 更新软件版本
  async updateVersion({ commit }, { versionId, versionData }) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.updateVersion(versionId, versionData)
      commit('UPDATE_VERSION', response.data)

      // 如果当前版本是更新后的版本，也更新当前版本
      if (state.currentVersion && state.currentVersion.id === response.data.id) {
        commit('SET_CURRENT_VERSION', response.data)
      }

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '更新软件版本失败')
      throw error
    }
  },

  // 删除软件版本
  async deleteVersion({ commit }, versionId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      await softwareApi.deleteVersion(versionId)
      commit('DELETE_VERSION', versionId)

      // 如果删除的是当前版本，清空当前版本
      if (state.currentVersion && state.currentVersion.id === versionId) {
        commit('SET_CURRENT_VERSION', null)
      }

      commit('SET_LOADING', false)
      return { success: true }
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '删除软件版本失败')
      throw error
    }
  },

  // 发布/下架软件版本
  async publishVersion({ commit }, { versionId, publish }) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.publishVersion(versionId, { publish })
      commit('UPDATE_VERSION', response.data)

      // 如果当前版本是更新后的版本，也更新当前版本
      if (state.currentVersion && state.currentVersion.id === response.data.id) {
        commit('SET_CURRENT_VERSION', response.data)
      }

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit(
        'SET_ERROR',
        error.response?.data?.message || (publish ? '发布软件版本失败' : '下架软件版本失败')
      )
      throw error
    }
  },

  // 下载软件版本
  async downloadVersion({ commit }, versionId) {
    try {
      commit('SET_LOADING', true)
      commit('CLEAR_ERROR')

      const response = await softwareApi.downloadVersion(versionId)

      commit('SET_LOADING', false)
      return response
    } catch (error) {
      commit('SET_LOADING', false)
      commit('SET_ERROR', error.response?.data?.message || '下载软件版本失败')
      throw error
    }
  },

  // 清空状态
  clearState({ commit }) {
    commit('CLEAR_STATE')
  },
}

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
}

File: backend/app/api/software.py
import os
from flask import Blueprint, request, jsonify, send_file, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from app.utils.response import success_response, error_response, admin_required
from app.utils.file import save_file, move_file_to_storage, delete_file, get_file_size_human_readable
from app.models.user import User
from app.models.software import SoftwareSpace, SoftwareVersion
from app.models.statistics import DownloadRecord
from app import db
from flask_sqlalchemy import SQLAlchemy

# 类型提示，避免pylance错误
from typing import Any, cast
session: Any = db.session

software_bp = Blueprint('software', __name__)


@software_bp.route('/software', methods=['GET'])
@jwt_required()
@admin_required
def get_spaces():
    """获取软件空间列表"""
    spaces = SoftwareSpace.query.order_by(SoftwareSpace.created_at.desc()).all()
    spaces_data = [space.to_dict() for space in spaces]
    
    return success_response(spaces_data)


@software_bp.route('/software', methods=['POST'])
@jwt_required()
@admin_required
def create_space():
    """创建软件空间"""
    data = request.get_json()
    
    if not data:
        return error_response("请求参数不能为空", 400)
    
    name = data.get('name')
    description = data.get('description')
    author = data.get('author')
    webhook_url = data.get('webhook_url')
    
    if not name:
        return error_response("软件名称不能为空", 400)
    
    current_user_id = get_jwt_identity()
    
    # 检查名称是否已存在
    existing_space = SoftwareSpace.query.filter_by(name=name).first()
    if existing_space:
        return error_response("软件名称已存在", 400)
    
    # 创建软件空间
    space = SoftwareSpace(
        name=name,
        description=description,
        author=author,
        webhook_url=webhook_url,
        created_by=current_user_id
    )
    
    session.add(space)
    session.commit()
    
    return success_response(space.to_dict(include_api_key=True), "软件空间创建成功")


@software_bp.route('/software/<int:space_id>', methods=['GET'])
@jwt_required()
@admin_required
def get_space(space_id):
    """获取软件空间详情"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    return success_response(space.to_dict(include_api_key=True))


@software_bp.route('/software/<int:space_id>', methods=['PUT'])
@jwt_required()
@admin_required
def update_space(space_id):
    """更新软件空间"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    data = request.get_json()
    
    if not data:
        return error_response("请求参数不能为空", 400)
    
    # 更新字段
    if 'name' in data:
        # 检查名称是否已存在（排除当前空间）
        existing_space = SoftwareSpace.query.filter(
            SoftwareSpace.name == data['name'],
            SoftwareSpace.id != space_id
        ).first()
        
        if existing_space:
            return error_response("软件名称已存在", 400)
        
        space.name = data['name']
    
    if 'description' in data:
        space.description = data['description']
    
    if 'author' in data:
        space.author = data['author']
    
    if 'webhook_url' in data:
        space.webhook_url = data['webhook_url']
    
    session.commit()
    
    return success_response(space.to_dict(include_api_key=True), "软件空间更新成功")


@software_bp.route('/software/<int:space_id>', methods=['DELETE'])
@jwt_required()
@admin_required
def delete_space(space_id):
    """删除软件空间"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    # 删除相关文件
    for version in space.versions:
        if os.path.exists(version.file_path):
            delete_file(version.file_path)
    
    session.delete(space)
    session.commit()
    
    return success_response(message="软件空间删除成功")


@software_bp.route('/software/<int:space_id>/regenerate-api-key', methods=['POST'])
@jwt_required()
@admin_required
def regenerate_api_key(space_id):
    """重新生成API密钥"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    new_api_key = space.regenerate_api_key()
    session.commit()
    
    return success_response({'api_key': new_api_key}, "API密钥重新生成成功")


@software_bp.route('/software/<int:space_id>/versions', methods=['GET'])
@jwt_required()
@admin_required
def get_versions(space_id):
    """获取软件版本列表"""
    space = SoftwareSpace.query.get_or_404(space_id)
    versions = space.versions.order_by(SoftwareVersion.created_at.desc()).all()
    
    versions_data = []
    for version in versions:
        version_dict = version.to_dict(include_file_path=True)
        version_dict['file_size_human'] = get_file_size_human_readable(version.file_size or 0)
        versions_data.append(version_dict)
    
    return success_response(versions_data)


@software_bp.route('/software/<int:space_id>/versions', methods=['POST'])
@jwt_required()
@admin_required
def create_version(space_id):
    """创建软件版本"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    # 检查是否有文件上传
    if 'file' not in request.files:
        return error_response("没有上传文件", 400)
    
    file = request.files['file']
    if file.filename == '':
        return error_response("没有选择文件", 400)
    
    # 获取其他参数
    version = request.form.get('version')
    release_note = request.form.get('release_note')
    documentation_url = request.form.get('documentation_url')
    
    if not version:
        return error_response("版本号不能为空", 400)
    
    # 检查版本是否已存在
    existing_version = SoftwareVersion.query.filter_by(
        space_id=space_id,
        version=version
    ).first()
    
    if existing_version:
        return error_response(f"版本 {version} 已存在", 400)
    
    current_user_id = get_jwt_identity()
    
    # 保存文件
    file_info, message = save_file(file)
    if not file_info:
        return error_response(message, 400)
    
    # 将文件移动到软件存储目录
    file_path, message = move_file_to_storage(
        file_info['file_path'],
        space_id,
        version
    )
    
    if not file_path:
        return error_response(message, 500)
    
    # 创建软件版本
    software_version = SoftwareVersion(
        space_id=space_id,
        version=version,
        file_path=file_path,
        file_size=file_info['file_size'],
        file_hash=file_info['file_hash'],
        release_note=release_note,
        documentation_url=documentation_url,
        created_by=current_user_id
    )
    
    session.add(software_version)
    session.commit()
    
    version_data = software_version.to_dict(include_file_path=True)
    version_data['file_size_human'] = get_file_size_human_readable(software_version.file_size or 0)
    
    return success_response(version_data, "软件版本创建成功")


@software_bp.route('/versions/<int:version_id>', methods=['GET'])
@jwt_required()
@admin_required
def get_version(version_id):
    """获取软件版本详情"""
    version = SoftwareVersion.query.get_or_404(version_id)
    
    version_data = version.to_dict(include_file_path=True)
    version_data['file_size_human'] = get_file_size_human_readable(version.file_size or 0)
    
    return success_response(version_data)


@software_bp.route('/versions/<int:version_id>', methods=['PUT'])
@jwt_required()
@admin_required
def update_version(version_id):
    """更新软件版本"""
    version = SoftwareVersion.query.get_or_404(version_id)
    
    data = request.get_json()
    
    if not data:
        return error_response("请求参数不能为空", 400)
    
    # 更新字段
    if 'release_note' in data:
        version.release_note = data['release_note']
    
    if 'documentation_url' in data:
        version.documentation_url = data['documentation_url']
    
    session.commit()
    
    version_data = version.to_dict(include_file_path=True)
    version_data['file_size_human'] = get_file_size_human_readable(version.file_size or 0)
    
    return success_response(version_data, "软件版本更新成功")


@software_bp.route('/versions/<int:version_id>', methods=['DELETE'])
@jwt_required()
@admin_required
def delete_version(version_id):
    """删除软件版本"""
    version = SoftwareVersion.query.get_or_404(version_id)
    
    # 删除文件
    if os.path.exists(version.file_path):
        delete_file(version.file_path)
    
    session.delete(version)
    session.commit()
    
    return success_response(message="软件版本删除成功")


@software_bp.route('/versions/<int:version_id>/publish', methods=['PUT'])
@jwt_required()
@admin_required
def publish_version(version_id):
    """发布/下架软件版本"""
    version = SoftwareVersion.query.get_or_404(version_id)
    
    data = request.get_json()
    
    if not data or 'publish' not in data:
        return error_response("请求参数不能为空", 400)
    
    publish = data['publish']
    
    if publish:
        version.publish()
        message = "软件版本发布成功"
    else:
        version.unpublish()
        message = "软件版本下架成功"
    
    session.commit()
    
    version_data = version.to_dict(include_file_path=True)
    version_data['file_size_human'] = get_file_size_human_readable(version.file_size or 0)
    
    return success_response(version_data, message)


@software_bp.route('/download/<int:version_id>', methods=['GET'])
@jwt_required()
@admin_required
def download_version(version_id):
    """下载软件版本（管理员）"""
    version = SoftwareVersion.query.get_or_404(version_id)
    
    if not os.path.exists(version.file_path):
        return error_response("文件不存在", 404)
    
    # 记录下载
    record = DownloadRecord(
        version_id=version.id,
        space_id=version.space_id,
        ip_address=request.remote_addr,
        user_agent=request.headers.get('User-Agent')
    )
    
    session.add(record)
    session.commit()
    
    # 发送文件
    return send_file(
        version.file_path,
        as_attachment=True,
        download_name=os.path.basename(version.file_path)
    )


# 公开API接口
@software_bp.route('/public/spaces', methods=['GET'])
def get_public_spaces():
    """获取所有公开的软件空间列表（公开API）"""
    # 获取所有有已发布版本的软件空间
    spaces = SoftwareSpace.query.join(
        SoftwareVersion,
        SoftwareSpace.id == SoftwareVersion.space_id
    ).filter(
        SoftwareVersion.is_published == True
    ).distinct().order_by(SoftwareSpace.created_at.desc()).all()
    
    spaces_data = []
    for space in spaces:
        space_dict = space.to_dict()
        space_dict.pop('api_key', None)  # 不返回API密钥
        
        # 添加已发布版本计数
        published_versions_count = SoftwareVersion.query.filter_by(
            space_id=space.id,
            is_published=True
        ).count()
        space_dict['versions_count'] = published_versions_count
        
        # 获取最新发布的版本
        latest_version = SoftwareVersion.query.filter_by(
            space_id=space.id,
            is_published=True
        ).order_by(SoftwareVersion.created_at.desc()).first()
        
        if latest_version:
            space_dict['latest_version'] = latest_version.version
            space_dict['latest_publish_date'] = latest_version.publish_date.isoformat() if latest_version.publish_date else None
        
        spaces_data.append(space_dict)
    
    return success_response(spaces_data)


@software_bp.route('/public/space/<int:space_id>', methods=['GET'])
def get_space_info_by_id(space_id):
    """通过ID获取软件空间信息（公开API）"""
    space = SoftwareSpace.query.get_or_404(space_id)
    
    # 检查是否有已发布的版本
    has_published_versions = SoftwareVersion.query.filter_by(
        space_id=space_id,
        is_published=True
    ).first() is not None
    
    if not has_published_versions:
        return error_response("该软件空间没有已发布的版本", 404)
    
    space_data = space.to_dict()
    space_data.pop('api_key', None)  # 不返回API密钥
    
    # 获取最新发布的版本
    latest_version = SoftwareVersion.query.filter_by(
        space_id=space_id,
        is_published=True
    ).order_by(SoftwareVersion.created_at.desc()).first()
    
    if latest_version:
        space_data['latest_version'] = latest_version.version
        space_data['latest_publish_date'] = latest_version.publish_date.isoformat() if latest_version.publish_date else None
    
    return success_response(space_data)


@software_bp.route('/public/<api_key>', methods=['GET'])
def get_space_info(api_key):
    """获取软件空间信息（公开API）"""
    space = SoftwareSpace.query.filter_by(api_key=api_key).first()
    
    if not space:
        return error_response("API密钥无效", 404)
    
    space_data = space.to_dict()
    space_data.pop('api_key', None)  # 不返回API密钥
    
    return success_response(space_data)


@software_bp.route('/public/<api_key>/versions', methods=['GET'])
def get_versions_info(api_key):
    """获取软件版本列表（公开API）"""
    space = SoftwareSpace.query.filter_by(api_key=api_key).first()
    
    if not space:
        return error_response("API密钥无效", 404)
    
    # 只返回已发布的版本
    versions = space.versions.filter_by(is_published=True).order_by(SoftwareVersion.created_at.desc()).all()
    
    versions_data = []
    for version in versions:
        version_dict = version.to_dict()
        version_dict['file_size_human'] = get_file_size_human_readable(version.file_size or 0)
        versions_data.append(version_dict)
    
    return success_response(versions_data)


@software_bp.route('/public/<api_key>/download/<string:version>', methods=['GET'])
def download_version_public(api_key, version):
    """下载软件版本（公开API）"""
    space = SoftwareSpace.query.filter_by(api_key=api_key).first()
    
    if not space:
        return error_response("API密钥无效", 404)
    
    # 查找指定版本
    if version.lower() == 'latest':
        software_version = space.get_latest_version()
        if not software_version:
            return error_response("没有可用的版本", 404)
    else:
        software_version = SoftwareVersion.query.filter_by(
            space_id=space.id,
            version=version,
            is_published=True
        ).first()
    
    if not software_version:
        return error_response("版本不存在或未发布", 404)
    
    if not os.path.exists(software_version.file_path):
        return error_response("文件不存在", 404)
    
    # 记录下载
    record = DownloadRecord(
        version_id=software_version.id,
        space_id=space.id,
        ip_address=request.remote_addr,
        user_agent=request.headers.get('User-Agent')
    )
    
    session.add(record)
    session.commit()
    
    # 发送文件
    return send_file(
        software_version.file_path,
        as_attachment=True,
        download_name=os.path.basename(software_version.file_path)
    )